### PR TITLE
Allow Assigning to DeviceNDArrays

### PR DESCRIPTION
**Pull request by njwhite**
Supports numpy-compatible broadcasting of the RHS of the assignment. Needs fixes from #2667 for the `test_broadcast` test to pass.

CC @sklam - this is what my PRs so far have been heading towards :)

### CODE REVIEW

**Comment by njwhite** [comment]
@sklam this is ready for review now!

File: numba/types/npytypes.py [hunk 1]
```diff
@@ -338,6 +338,11 @@ def can_convert_to(self, typingctx, other):
                 and (self.aligned or not other.aligned)):
                 return Conversion.safe
 
+        if self.ndim == 1 and isinstance(other, Sequence):
```
Comment (sklam) [inline numba/types/npytypes.py]: I am worried about the implicit conversion of arrays into tuples.  I think it is safer to do this with explicit conversion by introducing a new function.  The implicit conversion will affect overload resolution.

File: numba/cuda/cudadrv/devicearray.py [hunk 2]
```diff
@@ -425,6 +425,106 @@ def _do_getitem(self, item, stream=0):
...
+        for i, (l, r) in enumerate(zip(lhs.shape, rhs.shape)):
+            if r != 1 and l != r:
+                raise ValueError("Can't copy sequence with size %d to array axis %d with dimension %d" % (
+                    r,
+                    i,
+                    l))
+
+        # (3) do the copy
+
+        from numba import cuda, int64
```
[the remark this answers is not in the document]
Comment (njwhite) [reply numba/cuda/cudadrv/devicearray.py]: Fixed

File: numba/cuda/cudadrv/devicearray.py [hunk 3]
```diff
@@ -425,6 +425,106 @@ def _do_getitem(self, item, stream=0):
...
+        from numba.cuda.cudadrv.driver import driver
```
[the remark this answers is not in the document]
Comment (njwhite) [reply numba/cuda/cudadrv/devicearray.py]: Fixed

File: numba/cuda/cudadrv/devicearray.py [hunk 4]
```diff
@@ -517,10 +617,14 @@ def auto_device(obj, stream=0, copy=True):
     if _driver.is_device_memory(obj):
         return obj, False
     else:
-        sentry_contiguous(obj)
         if isinstance(obj, np.void):
             devobj = from_record_like(obj, stream=stream)
         else:
+            obj = np.array(
```
[the remark this answers is not in the document]
Comment (njwhite) [reply numba/cuda/cudadrv/devicearray.py]: Done!

File: numba/cuda/cudadrv/devicearray.py [hunk 5]
```diff
@@ -425,6 +425,106 @@ def _do_getitem(self, item, stream=0):
...
+                dtype=int64)
+
+            for i in range(ndim - 1, -1, -1):
+                idx[0, i] = location % lhs.shape[i]
+                idx[1, i] = (location % lhs.shape[i]) * (rhs.shape[i] > 1)
+                location //= lhs.shape[i]
+
+            lhs[cast(idx[0])] = rhs[cast(idx[1])]
+
+        kernel[(n_elements + tpb -1) // tpb, tpb, stream](
```
[the remark this answers is not in the document]
Comment (njwhite) [reply numba/cuda/cudadrv/devicearray.py]: I didn't know about this, it's very useful!

File: numba/cuda/cudadrv/devicearray.py [hunk 6]
```diff
@@ -425,6 +425,106 @@ def _do_getitem(self, item, stream=0):
...
+        def cast(it):
+            """
+            We can't dynamically look up an element in lhs with an array
+            due to `FancyIndexer`: it interprets `a[[1, 2]] = b` as
+            `a[1, :] = b` and `a[2, :] = b`. To generate the right code we need
+            to pass in a `BaseTuple` as the index -- so cast it here.
+            """
+            return it
+
+        tpb = driver.get_device().MAX_THREADS_PER_BLOCK
```
[the remark this answers is not in the document]
Comment (njwhite) [reply numba/cuda/cudadrv/devicearray.py]: No longer necessary due to the below

**Comment by njwhite** [comment]
@sklam re arrays<->tuples: I made the cast a Sequence as a 1-D array is unambiguously a sequence (and so that should have straightforwards integration with overloading semantics). The fact that the implementation round-trips through a tuple is just because that's the most straightforwards (even if it's not the most efficient). It should probably just interpret the array in-place, although it's not obvious how to do that. Would that be an OK change?

**Comment by sklam** [comment]
The problem with the cast to Sequence is that there are multiple subclasses of Sequence (e.g. List) but the conversion logic is only converting to tuple.   

As an alternative, I created https://github.com/numba/numba/pull/2691 to provide `to_fixed_tuple(array:Array<T>, length:ConstLength) -> Tuple<T, ConstLength>` to convert an array to a tuple of known length.  I have lightly tested it on your branch and successfully replaced 

```python 
       @cuda.jit('UniTuple(i8, %d)(i8[:])' % lhs.ndim, device=True)
        def cast(it):
            return it
```

with 

```python
       from numba.unsafe.ndarray import to_fixed_tuple
       @cuda.jit(device=True)
        def cast(it):
            return to_fixed_tuple(it, length=ndim)
```

**Comment by sklam** [comment]
I added a dependency to #2691, which should simplify this PR.

**Comment by njwhite** [comment]
@sklam thanks - that makes sense but it’d be good to have a zero-copy version (not that mine is), rather than making a new array and copying the items one by one just to cast the data-type. Maybe you can just return the array’s data pointer if the array is contiguous and the element types don’t need casting?

**Comment by sklam** [comment]
> it’d be good to have a zero-copy version

I haven't check it yet but LLVM should optimize the tuple_setitem away.

**Comment by njwhite** [comment]
@sklam WFM :) are the next steps (1) you get #2691 reviewed and merged then (2) I merge it to my branch and make my PR use the new functionality then (3) ask you to re-review? Or is there anything else I can help with?

**Comment by sklam** [comment]
@njwhite, that's right.  I will be pushing the last touch to my PR soon.

**Comment by njwhite** [comment]
@sklam I've made my PR use your new intrinsic now - could you re-review? Thanks -

**Comment by sklam** [comment]
This looks great!  Thanks for your contribution.

**Pull request merged**